rust/connlib/clients/android/src/lib.rs

```diff
@@ -411,11 +415,18 @@ fn connect(
         handle,
     };
 
-    let session = Session::connect(
+    let (private_key, public_key) = keypair();
+    let login = LoginUrl::client(
         api_url.as_str(),
-        secret,
+        &secret,
         device_id,
         Some(device_name),
+        public_key.to_bytes(),
+    )?;
```

Comment:
With this PR, creating a new `Session` is slightly more verbose. In particular, one needs to explicitly create a keypair first. That is because `LoginUrl` lives in `phoenix-channel`. That is where it should live IMO because it is its sole consumer and the format defined by the URL is directly coupled to the use of `phoenix-channel`[^1].

However, I think `phoenix-channel` should _not_ depend on boringtun and key-types, thus you have to pass in 32-bytes which represent the public key.

Overall, I think this is an okay trade-off:

1. We are now unifying, how we construct `LoginUrl`s across clients, gateways and the relay.
2. The only fallible point of `Session` is now creation of the runtime. As we continue to chip away at simplifying session, `connect` might eventually become infallible.

[^1]: At a future point, we might even be able to tie the "login-topic" to the `LoginUrl` via the type system, i.e. you will automatically login as a client if you made a `LoginUrl::client`.

Reply:
> I like that this opens a possibility for clients to persist their private key to disk if we ever want TOFU semantics with an even stronger "We can't decrypt your traffic even if we want" claim after 1.0

Just as a note, I don't think there's a stronger claim to privacy in this case, since we already generate the key in-memory, if we wanted to maliciously leak the key we could do it either way by sending the private key to the control plane.

Reply:
Well it's not about leaking but MITMing. If the control plane lies to the client and gateway and instead of giving them each other's pubkeys, it gives them the pubkeys of a relay it controls, and also passes off the relay as STUN/TURN candidate, the relay could decrypt traffic.

AIUI if the end user does not take any responsibility for key management, some kind of MITM attack is always possible. Otherwise they are ultimately trusting someone else to not lie about the pubkey.

Reply:
I mean if we get the private key we can just decrypt all traffic, whether we get the key from disk or create it in memory doesn't change that.

Reply:
I mean if the client relies on the API / portal to learn the gateway's pubkey, the portal could lie about the pubkey and MITM the client's traffic. Right?

Reply:
Correct. In this case, the portal is the root of trust.

We could theoretically allow an organization to point their clients and gateways to another root of trust in the future for the asymmetric keypair generation and/or the symmetric key.

Any organization this paranoid will just eat the operational overhead to self-host, including the IdP.

Reply:
Yeah the portal would have to direct traffic to a relay that was also in on this

Reply:
> Yeah the portal would have to direct traffic to a relay that was also in on this

That is not up to the portal to decide though, we will form a direct connection if we can, even if we are given a relay from the portal.

Reply:
But if the client* doesn't know the IP:port of the gateway, can the control plane not say, "Hm you're both behind unbreakable NATs, use this IP:port combo (which I secretly control) and here is the pubkey for your peer (which I secretly generated), trust me"?

* And vice verse for gateway not knowing where the client is